### PR TITLE
PB-511: Django request context in logs - #minor

### PR DESCRIPTION
**Pull request by benschs**
Add django request context to logs by filter configuration and middleware. Middleware adds the filter with the current request to the log handlers. Any logs will have the request information also added to the log record.

### CODE REVIEW

File: logging_utilities/request_middleware.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        self.logging_filter = logging_filter
+
+
+class AddRequestToLogMiddleware():
+    """Middleware that adds a logging filter *DjangoAppendRequestFilter* to the request.
+    """
+
+    def __init__(self, get_response: Callable[[WSGIRequest], Any], root: str = ""):
+        self.root = root
+        self.get_response = get_response
```
Comment (ltshb) [inline logging_utilities/request_middleware.py:38]: Might be a good idea to call `root` `rootLogger` to make the code easier to read/understand

File: logging_utilities/request_middleware.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            for handler in logger.handlers:
+                handlers.add(handler)
+        return list(handlers)
+
+    def _find_handlers_with_filter(self, filter_cls: type) -> dict[logging.Handler, list[Any]]:
+        """Dict of handlers mapped to their filters.
+        Only include handlers that have at least one filter of type *filter_cls*.
+        """
+        result = {}
+        for logger in self._find_handlers():
```
Comment (ltshb) [inline logging_utilities/request_middleware.py:76]: Here should it not be `handler` instead of `logger`?

File: tests/test_request_middleware.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import json
```
Comment (ltshb) [inline tests/test_request_middleware.py:1]: Maybe rename  to `test_django_request_middleware.py`

File: README.md [hunk 4]
```diff
@@ -507,6 +510,34 @@ filters:
...
+    request_attributes:
+      - path
+      - method
+      - META.QUERY_STRING
+    always_add: True
+```
+
+| Parameter    | Type | Default | Description                                    |
+|--------------|------|---------|------------------------------------------------|
+| `attributes` | list | None    | All request attributes that match any of the dotted keys of the list will be added to the jsonifiable object. When `None` then no attributes are added (default behavior). |
```
[the remark this answers is not in the document]
Comment (benschs) [reply README.md]: I change between `request_attributes` and `attributes` during development. It should now be `attributes` consistently.

File: logging_utilities/filters/django_append_request.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        self.request = request
+        self.attributes = request_attributes if request_attributes else list()
+        self.always_add = always_add
+
+    def filter(self, record: LogRecord) -> bool:
+        request = self.request
+        for attr in self.attributes:
+            val = r_getattr(request, attr, "-")
+            if self.always_add or val != "-":
+                setattr(record, "request." + attr, val)
```
[the remark this answers is not in the document]
Comment (benschs) [reply logging_utilities/filters/django_append_request.py]: If there is no request, we only add the value "-" if `always_add` is explicitly set, so this is fine for me as it is.
I added a test case that explicitly sets `None` as request to demonstrate the behavior.

File: logging_utilities/filters/django_append_request.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            always_add: bool
+                Always add attributes even if they are missing. Missing attributes with have the
+                value "-".
+        """
+        self.request = request
+        self.attributes = attributes if attributes else list()
+        self.always_add = always_add
+
+    def filter(self, record: LogRecord) -> bool:
+        request = self.request
```
Comment (boecklic) [inline logging_utilities/filters/django_append_request.py:41]: in `__init__` the WSGIRequest is marked as optional and can be `None` if I understand correctly, so wouln't you need to check for `None` here and only continue if not None?

File: README.md [hunk 7]
```diff
@@ -507,6 +509,44 @@ filters:
 
 **NOTE**: `JsonDjangoRequest` only support the special key `'()'` factory in the configuration file (it doesn't work with the normal `'class'` key).
 
+## Django request log records
+
+To add context information from the current request to each log record. For this the filter `DjangoAppendRequestFilter` must be added as well as the middleware `AddRequestToLogMiddleware`.
```
Comment (boecklic) [inline README.md:514]: maybe you could be a little more verbose about the basic idea behind your solution with middleware and filter. With some not very deep understanding of how logging filters work this is not immediately obvious.

File: logging_utilities/django_middlewares/request_middleware.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    def __init__(self, get_response: Callable[[WSGIRequest], Any], root_logger: str = ""):
+        self.root_logger = root_logger
+        self.get_response = get_response
+
+    def __call__(self, request: WSGIRequest) -> Any:
+        w_request = WrappedRequest.from_parent(request, None)
+        response = self.process_request(w_request)
+        if not response:
+            response = self.get_response(w_request)
```
Comment (boecklic) [inline logging_utilities/django_middlewares/request_middleware.py:44]: I'm not sure I understand this part: the `process_request` method already does a `return self.get_response(request)`, so if that `return` doesn't evaluate to sth `true`, why should calling it again result in a different outcome?

File: logging_utilities/django_middlewares/request_middleware.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        response = self.process_response(w_request, response)
+
+        return response
+
+    def _find_loggers(self) -> dict[str, logging.Logger]:
+        """Return loggers part of root.
```
Comment (boecklic) [inline logging_utilities/django_middlewares/request_middleware.py:50]: a little more information about what you do here, i.e. what `root_logger` is about would be helpful. This is not directly obvious also from django middleware doc I think

**Review by boecklic — APPROVED** [review]
Nice! Do you have an idea about possible performance impact of this solution? Since it's doing the processing for every request.

**Review by ltshb — CHANGES_REQUESTED** [review]
Could you verify if the filter approach is working in a multithread environment ? As discussed yesterday the logging loggers instance and filters are globals and it might be an issue with mutlithread environement.

**Comment by benschs** [comment]
> Could you verify if the filter approach is working in a multithread environment ? As discussed yesterday the logging loggers instance and filters are globals and it might be an issue with mutlithread environement.

I could verify that this approach does not work in a multithreaded environment.
I will close this PR and open a new one with another approach of using thread variables.

**Pull request closed without merging**